src/charm.py

```diff
@@ -31,12 +31,53 @@ def __init__(self, *args):
 
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.mongodb_relation_joined, self._mongodb_relation_joined)
         self.framework.observe(self.on.start, self._on_start)
 
+    def _mongodb_relation_joined(self, event) -> None:
+        if not self.unit.is_leader():
+            return
+
+        # do not initialise replica set until all peers have joined
+        if not self.app.planned_units() == len(self._peers.units) + 1:
+            logger.debug(
+                "planned units does not match joined units: %s != %s, will "
+                + "wait to initialise replica set until all planned units "
+                + "have joined",
+                str(self.app.planned_units()),
+                len(self.model.get_relation(PEER).units) + 1,
+            )
+            return
+
+        # do not initilise replica set until all peers have started mongod
+        for peer in self._peers.units:
+            peer_ip = str(self._peers.data[peer].get("private-address"))
+            peer_config = self._config
+            peer_config["calling_unit_ip"] = peer_ip
+            mongo_peer = MongoDB(peer_config)
+            if not mongo_peer.is_ready(all_replicas=False):
+                logger.debug(
+                    "unit: %s is not ready, cannot initilise replica set "
+                    + "until all units are ready, defering on relation-joined",
+                    peer,
+                )
+                event.defer()
```

Comment:
So what `self._mongo.is_ready(all_replicas=True)` does is it checks if all the replicas in the replica set are available. (For this to return true the replica set **must** at first be initialised). Until it is, each replica will need to be checked individually which is why we check them in this way. Does that make sense?

You asking this makes me think I should change the variable `all_replicas` would that make things more clear? Perhaps something like `as_replica_set: bool` or `as_stand_alone: bool`?  I think `as_stand_alone` would make the most sense here, so it would be: `if not mongo_peer.is_ready(as_stand_alone=True):`

Reply:
@marceloneppel